### ISSUE TITLE
goroutine leaks from beginSegment

### ISSUE THREAD

**Issue by elan-sg** [aws/aws-xray-sdk-go#51]
https://github.com/aws/aws-xray-sdk-go/blob/8d5ff7cd4d96b6e380cf7bcaf576ed1848b10b30/xray/segment.go#L70-L75

This goroutine will leak if a context without cancel is used to beginSegment, I think Segment.Close should clean this up.

**Comment by haotianw465** [comment]
Hi, 

Sorry for the late response. Thank you for reporting the issue. We are aware of this and will have a fix soon. Please stay tuned.

**Comment by Cyberax** [comment]
Guys, it's been a year so far. The fix is simple, just add a select{} statement that monitors a channel that is closed in segment.Close().

**Comment by yogiraj07** [comment]
Hello @Cyberax ,
Apologies for delay in solving this issue. I am happy to work with you and prioritize on solving it.  Can you post a PR for the above commit with unit tests. 

Best,
Yogi

**Comment by LeoAdamek** [comment]
Any update on this? It's quite a big issue as, even when I explicitly complete the context passed to the segment, there are still goroutines leaked. In an API the rate these leaks occur can be up to the every request.

In my case I've written a minimal middleware for Gin to handle tracing:

~~~~go
g.Use(func(c *gin.Context) {
		ctx := context.WithValue(c, ctxURL, c.Request.URL.String())

		serviceName := "solos-api"

		if envName := viper.GetString("xray.name"); envName != "" {
			serviceName += " " + envName
		}

		parts := strings.Split(c.HandlerName(), ".")

		serviceName += " " + parts[len(parts)-1]

		ctx, s := xray.BeginSegment(ctx, serviceName)

		remote := c.Request.Header.Get("X-Forwarded-For")

		if remote == "" {
			remote = c.Request.RemoteAddr
		}

		s.HTTP = &xray.HTTPData{}
		s.HTTP.Request = &xray.RequestData{
			Method:    c.Request.Method,
			ClientIP:  remote,
			URL:       c.Request.URL.Path,
			UserAgent: c.Request.UserAgent(),
		}

		// Add the trace Id as a header so the request can be referenced later
		// for troubleshooting.
		c.Writer.Header().Set("X-Trace-Id", s.TraceID)

		c.Set("ctx", ctx)
		c.Next()
		status := c.Writer.Status()
		s.HTTP.Response = &xray.ResponseData{
			Status: status,
		}

		s.Error = false
		s.Fault = false

		if status >= 400 {
			s.Error = true
			if status >= 500 {
				s.Fault = true
			}
		}

		ctx.Done()
		s.Close(nil)
	})
~~~~

However, no matter what I try, there are still goroutines leaked for almost every request.

**Comment by dtarico-suplari** [comment]
Just had this impact a production deployment after we added xray.  This but has been open a year and a half and there's neither a fix nor a big warning in the docs.  Not cool.

**Comment by JakeKalstad** [comment]
Is there any foreseeable action on this issue? We're seeing massive memory leaks resulting in crashes in production upon release. The documentation for `Start a custom segment/subsegment` is either incorrect or incomplete as we are properly closing the segment and still seeing issues. After migrating from a background context to a `context.WithCancel()` we are still seeing the memory leak.

Thanks.

**Comment by bhautikpip** [comment]
Hi @JakeKalstad,

Apologies for the inconvenience. We recently merged PR (https://github.com/aws/aws-xray-sdk-go/pull/156). This might fix some issues. Feel free to submit a PR or open an issue if you find any other source of memory leaks.

**Comment by JakeKalstad** [comment]
Thanks for the quick response, we've confirmed that it has resolved our issue.

**Comment by michelaquino** [comment]
Hi @bhautikpip 
Was this fix already released? The last version is `v1.0.0-rc.14` and do not contains the fix.

**Comment by bhautikpip** [comment]
Hi @michelaquino,

This fix is yet to be released. This fix will not be in (v1.0.0-rc.14) this version.

**Comment by bhautikpip** [comment]
Hi @michelaquino,

Just to keep you updated we are planning a release including this fix soon.

**Comment by bhautikpip** [comment]
I am closing this issue since we have pushed out a fix for this issue in this release (https://github.com/aws/aws-xray-sdk-go/releases/tag/v1.0.0-rc.15). Feel free to open another issue if you think it's not addressed.

**Comment by zapling** [comment]
This issue is not fixed in `v1.7.0`, we ran into this in services when we upgraded xray. Should we open a new ticket for this? @bhautikpip
